### PR TITLE
[C/PyTorch] Add FP8 DPA and MHA

### PR DESCRIPTION
**Pull request by cyanguwa**
This PR adds support for FP8 dot product attention (DPA) and multi-head attention (MHA).

FP8 DPA:
1. To enable: add `fp8_dpa=True` to FP8 recipe, and place model under `with fp8_autocast(enabled=True):`.
2. Example workflow: `LayerNormLinear` -> `DPA (cast BF16 input to FP8, FP8 DPA, cast output to BF16)` -> `Linear`
3. Train with FP8 DPA in fprop only: add `export NVTE_FP8_DPA_BWD=0` to disable FP8 DPA bprop

FP8 MHA:
1. To enable: add `fp8_mha=True, fp8_dpa=True` to FP8 recipe, and place model under `with fp8_autocast(enabled=True):`.
2. Example workflow: `LayerNormLinear (output in FP8)` -> `DPA (FP8 DPA)` -> `(FP8 input) Linear`
3. Train with FP8 DPA in fprop only: add `export NVTE_FP8_DPA_BWD=0` to disable FP8 DPA bprop
4. FP8 MHA avoids the casting ops at the module boundaries compared to FP8 DPA. 

This PR also changes the `FusedAttnBackend["F16_arbitrary_seqlen"]` and `FusedAttnBackend["FP8"]` backends to always output `stats` (softmax statistic tensor) to facilitate with eval and context parallel use cases.

### CODE REVIEW

File: transformer_engine/common/fused_attn/fused_attn.cpp [hunk 1]
```diff
@@ -85,15 +85,26 @@ NVTE_Fused_Attn_Backend nvte_get_fused_attn_backend(
...
-  if ((q_dtype == NVTEDType::kNVTEFloat8E4M3) || (q_dtype == NVTEDType::kNVTEFloat8E5M2)
-          && (sm_arch_ >= 90)
-          && (max_seqlen_q == max_seqlen_kv)
-          && (num_attn_heads == num_gqa_groups)
-          && (max_seqlen_q <= 512)
-          && (head_dim == 64)
-          && (bias_type == NVTE_Bias_Type::NVTE_NO_BIAS)
-          && (attn_mask_type == NVTE_Mask_Type::NVTE_PADDING_MASK)
-          && (qkv_layout == NVTE_QKV_Layout::NVTE_T3HD)) {
+  if (((q_dtype == NVTEDType::kNVTEByte)
```
Comment (ptrendx) [inline transformer_engine/common/fused_attn/fused_attn.cpp:88]: Why add the Byte type here?

Comment (cyanguwa) [reply transformer_engine/common/fused_attn/fused_attn.cpp:88]: We don't have the fp8 dtypes here: https://github.com/cyanguwa/TransformerEngine/blob/01685ae25f7c2834bfbfcfe3b9e50f42dc2510a6/transformer_engine/pytorch/attention.py#L3488

Comment (cyanguwa) [reply transformer_engine/common/fused_attn/fused_attn.cpp:88]: Is there a way to access `fp8_meta`? DPA currently is not a `TEBaseModule`.

File: transformer_engine/pytorch/float8_tensor.py [hunk 2]
```diff
@@ -206,6 +206,87 @@ def forward(
     def backward(ctx, grad):
         return grad.to(ctx.input_dtype), None
 
+class _ViewFunc(torch.autograd.Function):
```
Comment (yaox12) [inline transformer_engine/pytorch/float8_tensor.py:209]: @timmoon10 Do we prefer adding a `_ViewFunc` or registering `aten.view.default` in `__torch_dispatch__`?

File: transformer_engine/common/recipe/__init__.py [hunk 3]
```diff
@@ -126,6 +126,8 @@ def scaling_factor_compute(amax: Tensor,
     override_linear_precision: _OverrideLinearPrecision = _OverrideLinearPrecision()
     scaling_factor_compute_algo: Optional[Callable] = None
     reduce_amax: bool = True
+    fp8_dpa: bool = False
+    fp8_mha: bool = False
```
Comment (ksivaman) [inline transformer_engine/common/recipe/__init__.py:130]: Missing docs

File: transformer_engine/common/fused_attn/fused_attn.cpp (hunk 1 above)
Comment (ptrendx) [reply transformer_engine/common/fused_attn/fused_attn.cpp:88]: There is at least `_fp8_dtype` property of the Float8Tensor you could access (not sure why we don't expose it as a regular property without `_`, @ksivaman any insight?). Then you would be able to do `query_layer._dp8_dtype` instead.

File: transformer_engine/common/fused_attn/fused_attn_f16_arbitrary_seqlen.cu [hunk 4]
```diff
@@ -147,7 +147,7 @@ void fused_attn_arbitrary_seqlen_fwd_impl(
             fe::graph::SDPA_attributes sdpa_options;
             sdpa_options = fe::graph::SDPA_attributes()
                             .set_name("flash_attention")
-                            .set_is_inference(!is_training)
+                            .set_is_inference(false)
```
Comment (ptrendx) [inline transformer_engine/common/fused_attn/fused_attn_f16_arbitrary_seqlen.cu:150]: Why?

Comment (ptrendx) [reply transformer_engine/common/fused_attn/fused_attn_f16_arbitrary_seqlen.cu:150]: Is it because of CP?

File: transformer_engine/common/recipe/__init__.py [hunk 5]
```diff
@@ -106,6 +106,20 @@ def scaling_factor_compute(amax: Tensor,
                 GPU maintains local amaxes and scaling factors. To ensure results are
                 numerically identical across checkpointing boundaries in this case, all
                 ranks must checkpoint in order to store the local tensors.
+    fp8_dpa: bool, default = `False`
```
Comment (ptrendx) [inline transformer_engine/common/recipe/__init__.py:109]: Please mark this and fp8_mha options as BETA functionality

File: transformer_engine/common/fused_attn/fused_attn_f16_arbitrary_seqlen.cu (hunk 4 above)
Comment (cyanguwa) [reply transformer_engine/common/fused_attn/fused_attn_f16_arbitrary_seqlen.cu:150]: Yes

File: transformer_engine/common/fused_attn/fused_attn.cpp (hunk 1 above)
Comment (cyanguwa) [reply transformer_engine/common/fused_attn/fused_attn.cpp]: Fixed. Using `_fp8_dtype` now.

File: transformer_engine/common/recipe/__init__.py (hunk 5 above)
Comment (cyanguwa) [reply transformer_engine/common/recipe/__init__.py:109]: Done. Let me know if the language needs tweaking.

File: transformer_engine/pytorch/float8_tensor.py (hunk 2 above)
Comment (cyanguwa) [reply transformer_engine/pytorch/float8_tensor.py:236]: `view()` and `reshape()` here produce `Float8Tensor`s rather than `torch.Tensor`s, and it has proven helpful when passing `Float8Tensor`s around. Let me know if there's a better way to do it though.

**Review by ksivaman — APPROVED** [review]
Unblocking for 1.6. For future work we should consider a refactor that doesn't require non-parameter modules (like DPA) to be base modules.

File: transformer_engine/pytorch/module/linear.py [hunk 6]
```diff
@@ -338,6 +381,10 @@ def forward(
     def backward(
         ctx, grad_output: torch.Tensor
     ) -> Tuple[Union[torch.Tensor, None], ...]:
+        if isinstance(grad_output[0], Float8Tensor):
```
Comment (Victarry) [inline transformer_engine/pytorch/module/linear.py:384]: Hi, @cyanguwa, this should be `grad_output` since `grad_output` is a Tensor instead of list.

File: transformer_engine/pytorch/module/base.py [hunk 7]
```diff
@@ -684,16 +694,22 @@ def grad_output_preprocess(
...
+            if not isinstance(grad_output_mat, Float8Tensor):
+                cast_to_fp8(
+                    grad_output_mat,
+                    ctx.fp8_meta["scaling_bwd"],
+                    tex.FP8BwdTensors.GRAD_OUTPUT1,
+                    fp8_dtype_backward,
+                    out=grad_output_c,
+                )
+            else:
+                grad_output_c = grad_ouput_mat # pylint: disable=undefined-variable
```
Comment (yaox12) [inline transformer_engine/pytorch/module/base.py:706]: @cyanguwa This is a typo. It should be `grad_output_mat`.

Comment (cliffwoolley) [reply transformer_engine/pytorch/module/base.py:706]: good catch - we will fix it.

Comment (timmoon10) [reply transformer_engine/pytorch/module/base.py:706]: https://github.com/NVIDIA/TransformerEngine/pull/795
